File: src/pressure_to_pose.cpp
#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/fluid_pressure.hpp>

#include <algorithm>
#include <cmath>
#include <memory>
#include <string>

namespace
{

class PressureToPose : public rclcpp::Node
{
public:
  PressureToPose()
  : Node("pressure_to_pose")
  {
    declare_parameter<std::string>("input_topic", "sensors/pressure");
    declare_parameter<std::string>("output_topic", "sensors/pressure/pose");
    declare_parameter<std::string>("environment", "real");
    declare_parameter<std::string>("frame_id", "world_enu");
    declare_parameter<std::string>("sensor_frame_id", "");
    declare_parameter<bool>("positive_down", true);
    declare_parameter<double>("surface_pressure_pa", 101325.0);
    declare_parameter<double>("fluid_density_kg_m3", 997.0);
    declare_parameter<double>("gravity_m_s2", 9.80665);
    declare_parameter<double>("z_scale", 1.0);
    declare_parameter<double>("z_offset_m", 0.0);
    declare_parameter<double>("fallback_z_variance", 0.01);
    declare_parameter<double>("fallback_xy_variance", 99999.0);

    const auto input_topic = get_parameter("input_topic").as_string();
    const auto output_topic = get_parameter("output_topic").as_string();

    publisher_ = create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(
      output_topic,
      10);
    subscription_ = create_subscription<sensor_msgs::msg::FluidPressure>(
      input_topic,
      10,
      [this](const sensor_msgs::msg::FluidPressure::SharedPtr msg) {
        on_pressure(*msg);
      });

    RCLCPP_INFO(
      get_logger(),
      "Converting pressure %s to pose %s",
      input_topic.c_str(),
      output_topic.c_str());
  }

private:
  void on_pressure(const sensor_msgs::msg::FluidPressure & msg)
  {
    const double density = get_parameter("fluid_density_kg_m3").as_double();
    const double gravity = get_parameter("gravity_m_s2").as_double();
    if (density <= 0.0 || gravity <= 0.0) {
      RCLCPP_WARN_THROTTLE(
        get_logger(),
        *get_clock(),
        2000,
        "Invalid density or gravity; skipping pressure sample");
      return;
    }

    const auto environment = get_parameter("environment").as_string();
    double depth = 0.0;

    if (environment == "sim") {
      if (!has_sim_surface_pressure_) {
        sim_surface_pressure_ = msg.fluid_pressure;
        has_sim_surface_pressure_ = true;
        RCLCPP_INFO(
          get_logger(),
          "Using %.6f as simulated surface pressure reference",
          sim_surface_pressure_);
      }

      const double pressure_delta_pa = msg.fluid_pressure - sim_surface_pressure_;
      depth = pressure_delta_pa / (density * gravity);
    } else {
      const double surface_pressure = get_parameter("surface_pressure_pa").as_double();
      const double pressure_delta_pa = msg.fluid_pressure - surface_pressure;
      depth = pressure_delta_pa / (density * gravity);
    }

    if (!std::isfinite(depth)) {
      return;
    }

    const bool positive_down = get_parameter("positive_down").as_bool();
    const double raw_z = positive_down ? -depth : depth;
    const double z = raw_z * get_parameter("z_scale").as_double() +
      get_parameter("z_offset_m").as_double();

    geometry_msgs::msg::PoseWithCovarianceStamped pose;
    pose.header.stamp = msg.header.stamp;
    pose.header.frame_id = get_parameter("frame_id").as_string();
    pose.pose.pose.position.z = z;
    pose.pose.pose.orientation.w = 1.0;

    double z_variance = get_parameter("fallback_z_variance").as_double();
    if (msg.variance > 0.0 && std::isfinite(msg.variance)) {
      z_variance = msg.variance / ((density * gravity) * (density * gravity));
    }

    const double xy_variance = get_parameter("fallback_xy_variance").as_double();
    pose.pose.covariance[0] = std::max(0.0, xy_variance);
    pose.pose.covariance[7] = std::max(0.0, xy_variance);
    pose.pose.covariance[14] = std::max(0.0, z_variance);
    pose.pose.covariance[21] = 99999.0;
    pose.pose.covariance[28] = 99999.0;
    pose.pose.covariance[35] = 99999.0;

    publisher_->publish(pose);
  }

  rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr publisher_;
  rclcpp::Subscription<sensor_msgs::msg::FluidPressure>::SharedPtr subscription_;

  bool has_sim_surface_pressure_{false};
  double sim_surface_pressure_{0.0};
};

}  // namespace

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PressureToPose>());
  rclcpp::shutdown();
  return 0;
}

File: sura_localization/pressure_to_pose.py
#!/usr/bin/env python3

import math

import rclpy
from geometry_msgs.msg import PoseWithCovarianceStamped
from rclpy.node import Node
from sensor_msgs.msg import FluidPressure


class PressureToPose(Node):
    def __init__(self):
        super().__init__("pressure_to_pose")

        self.declare_parameter("input_topic", "/cirtesub/sensors/pressure")
        self.declare_parameter("output_topic", "/cirtesub/sensors/pressure/pose")
        self.declare_parameter("frame_id", "odom")
        self.declare_parameter("sensor_frame_id", "cirtesub/Pressure")
        self.declare_parameter("fluid_density", 1000.0)
        self.declare_parameter("gravity", 9.80665)
        self.declare_parameter("pressure_offset_pa", 101325.0)
        self.declare_parameter("positive_down", False)
        self.declare_parameter("fallback_z_variance", 0.01)
        self.declare_parameter("unused_axis_variance", 1000000.0)

        input_topic = self.get_parameter("input_topic").value
        output_topic = self.get_parameter("output_topic").value

        self._publisher = self.create_publisher(
            PoseWithCovarianceStamped,
            output_topic,
            10,
        )
        self.create_subscription(FluidPressure, input_topic, self._pressure_callback, 10)

    def _pressure_callback(self, msg: FluidPressure) -> None:
        density = float(self.get_parameter("fluid_density").value)
        gravity = float(self.get_parameter("gravity").value)
        if density <= 0.0 or gravity <= 0.0:
            self.get_logger().warn("Fluid density and gravity must be positive.")
            return

        pressure_offset_pa = float(self.get_parameter("pressure_offset_pa").value)
        depth = (msg.fluid_pressure - pressure_offset_pa) / (density * gravity)
        z = -depth if bool(self.get_parameter("positive_down").value) else depth

        pose = PoseWithCovarianceStamped()
        pose.header.stamp = msg.header.stamp
        pose.header.frame_id = str(self.get_parameter("frame_id").value)
        pose.pose.pose.position.z = z
        pose.pose.pose.orientation.w = 1.0

        unused_variance = float(self.get_parameter("unused_axis_variance").value)
        covariance = [0.0] * 36
        for index in range(6):
            covariance[index * 6 + index] = unused_variance

        pressure_variance = msg.variance
        if pressure_variance > 0.0 and math.isfinite(pressure_variance):
            z_variance = pressure_variance / ((density * gravity) ** 2)
        else:
            z_variance = float(self.get_parameter("fallback_z_variance").value)

        covariance[14] = z_variance
        pose.pose.covariance = covariance

        self._publisher.publish(pose)


def main(args=None):
    rclpy.init(args=args)
    node = PressureToPose()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
